File: data_structure/graph/拓扑排序/0207.cpp
#include <iostream>
#include <vector>
#include <array>
#include <queue>
using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses, vector<int>(numCourses, 0));
        vector<int> indegree(numCourses, 0);

        for (int i = 0; i < prerequisites.size(); ++i) {
            int start = prerequisites[i][1];
            int end = prerequisites[i][0];
            graph[start][end] = 1;
            indegree[end] += 1;
        }

        queue<int> s;
        for (int i = 0; i < numCourses; ++i) {
            if (indegree[i] == 0) {
                s.push(i);
            }
        }

        vector<int> rst;
        while (!s.empty()) {
            int n = s.front();
            s.pop();
            rst.push_back(n);
            for (int i = 0; i < numCourses; ++i) {
                if (graph[n][i] != 0) {
                    graph[n][i] = 0;
                    indegree[i]--;
                    if (indegree[i] == 0) {
                        s.push(i);
                    }
                }
            }
        }

        if (rst.size() == numCourses) {
            return true;
        }
        return false;
    }
};


int main() {
    cout << "hello" << endl;
}

File: data_structure/graph/拓扑排序/0207.py
"""
这题用拓扑排序的解法
"""

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[0 for i in range(numCourses)] for i in range(numCourses)]
        indegree = [0 for i in range(numCourses)]

        for item in prerequisites:
            start = item[1]
            end = item[0]
            graph[start][end] = 1
            indegree[end] += 1
        
        s = set()
        for i in range(numCourses):
            if indegree[i] == 0:
                s.add(i)
        
        rst = []
        while len(s) != 0:
            n = s.pop()
            rst.append(n)
            for i in range(numCourses):
                if graph[n][i] != 0:
                    graph[n][i] = 0
                    indegree[i] -= 1
                    if indegree[i] == 0:
                        s.add(i)
        
        if len(rst) != numCourses:
            return False
        return True
